WordPress/src/main/java/org/wordpress/android/ui/media/MediaSettingsActivity.java

```diff
@@ -297,8 +297,8 @@ public void handleOnBackPressed() {
             int mScrollRange = -1;
 
             @Override
-            public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
-                if (mScrollRange == -1) {
+            public void onOffsetChanged(@Nullable AppBarLayout appBarLayout, int verticalOffset) {
+                if (mScrollRange == -1 && appBarLayout != null) {
```

Comment:
Great question and my short answer here is; I am not sure. I didn't want to invest more time into going deeper than than. However, I think this question of yours about an invalid state makes a case about being in favor of the "fail fast" case. Otherwise, going the "fail silently" road, we might introduce invalid states that are going to be much harder to identify going forward. 🤷 🤔

Reply:
I think the only "fail silently" part about this is the addition of the `appBarLayout != null`, which is a quick way to make the compile error go away, but not necessarily a "valid" thing to do. To me, I think of `@Nullable` as failing at _compile time_, and `@NonNull` as failing at _runtime_, and thus `@Nullable` is failing _faster_, since it fails before even reaching users.

I think the conflation here is that _additional changes_ are being introduced, _alongside_ the application of the `@Nullable` annotation, to satisfy the compiler (or make the red things go away in the IDE 😄). But, in many cases, significant effort is required to do this correctly, as well as a deeper understanding of the implementation in question.

This is another reason I think these kinds of changes should be postponed to a different PR. The effort to review these may vary greatly from case to case, and it would be ideal to have input from folks more familiar with a given implementation (especially regarding the correct handing of the `@Nullable` cases).

Reply:
> To me, I think of @Nullable as failing at compile time, and @NonNull as failing at runtime, and thus @Nullable is failing faster, since it fails before even reaching users.

Hmmm... yea @mkevins , I get that but let me provide an extra thing that `@Nullable` is to me. In addition to it failing at compile time, and thus failing "faster", if used incorrectly, it is now actually failing "silently" as now we are introducing a whole application flow and could never be valid, or an application state that is ignored, which, in its turn will propagate this downstream everywhere else. Plus, doing that, we are now giving the wrong example for others to follow.

Thus, to me, using `@NonNull` is already falling "faster" vs using `@Nullable` as I consider that you could never fail with that anyway, you could only "fail" during compile time, but I personally don't consider that failing, I only consider failing a runtime thing. 🤔

Reply:
It seems I should clarify what I mean by `@Nullable` failing. I don't mean that it is failing in the sense that we add dead code unnecessarily. I mean that the _build_ will fail, because the compiler will not _let_ us dereference a pointer to memory that is annotated this way.

> I only consider failing a runtime thing

To me, I have a strong preference that as many failures (or errors, etc.) are surfaced at compile time as possible. The preference of where a "failure" should be surfaced is: [runtime] < [compile time] < [lint].

> if used incorrectly, it is now actually failing "silently" as now we are introducing a whole application flow and could never be valid, or an application state that is ignored, which, in its turn will propagate this downstream everywhere else. Plus, doing that, we are now giving the wrong example for others to follow.

I don't think having wrongly applied `@Nullable` annotations would lead to an application state which is ignored or invalid, but rather that it may make the compile complain until non-existent cases are handled (i.e. the compile will force the writing of dead code). I'm also not worried for the propagation of this because we have things like `!!` and `Object.requireNotNull` (in Kotlin and Java, respectively), which give us the power to decide how far along the codepaths we want to provide null-safety for an uncertain nullness. We do _not_ have an equivalent mechanism for `@NonNull` annotations. Once annotated as such, the NPE will occur _anywhere_ this wrong assumption leads. I.e., user crashes that are trickier to track down, because the originally incorrect annotation is not the site of the crash.

Another point to consider is that "propagation" is going to happen with _either_ of the annotations. For example, annotating a _return type_ as `@Nullable` requires any instances, fields, other return types, etc. that consume such a method be annotated similarly, while annotating a _parameter_ with `@NonNull` similarly requires at the call site that any arguments passed are non-null. In _either case_, in order for the code to remain _safe_ and _valid_, the developer would need to provide some handling of the potentially null case before usage, with the choices being primarily conditionally / gracefully failing, or asserting and crashing. It is, of course, _easier_ to just crash, but users tend not to like that 😅 . But handling these cases gracefully requires an understanding of the domain - a scope far to large to handle in bulk, imo.